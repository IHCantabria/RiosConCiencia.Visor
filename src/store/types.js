//DATAHUB

//Mutations
export const M_SET_PRODUCTS = "datahub/M_SET_PRODUCTS";
export const M_SET_SOURCES = "datahub/M_SET_SOURCES";
export const M_SET_PRODUCTMOMENTS = "datahub/M_SET_PRODUCTMOMENTS";
export const M_SET_PRODUCTVARIABLETYPES = "datahub/M_SET_PRODUCTVARIABLETYPES";
export const M_UPDATE_PRODUCTS_SELECTED = "datahub/M_UPDATE_PRODUCTS_SELECTED";

//Actions
export const A_FETCH_DATAHUB_PRODUCTS = "datahub/A_FETCH_DATAHUB_PRODUCTS";
export const A_FETCH_DATAHUB_SOURCES = "datahub/A_FETCH_DATAHUB_SOURCES";
export const A_FETCH_DATAHUB_PRODUCTMOMENTS =
  "datahub/M_SET_DATAHUB_PRODUCTMOMENTS";
export const A_FETCH_DATAHUB_PRODUCTVARIABLETYPES =
  "datahub/A_FETCH_DATAHUB_PRODUCTVARIABLETYPES";
export const A_SET_CLEAR_PRODUCTS = "datahub/A_SET_CLEAR_PRODUCTS";
export const A_UPDATE_PRODUCTS_SELECTED = "datahub/A_UPDATE_PRODUCTS_SELECTED";

//Getters
export const G_GET_PRODUCTS_SORTED = "datahub/G_GET_PRODUCTS_SORTED";
export const G_GET_PRODUCTS_SELECTED = "datahub/G_GET_PRODUCTS_SELECTED";
export const G_GET_PRODUCTS_COUNT = "datahub/G_GET_PRODUCTS_COUNT";
export const G_GET_PRODUCT_BY_MAPID = "datahub/G_GET_PRODUCT_BY_MAPID";
export const G_GET_SOURCE_NAMESHORT = "datahub/G_GET_SOURCE_NAMESHORT";
export const G_GET_SOURCE_TYPE = "datahub/G_GET_SOURCE_TYPE";
export const G_GET_PRODUCTMOMENTS_SORTED =
  "datahub/G_GET_PRODUCTMOMENTS_SORTED";
export const G_GET_PRODUCTVARIABLETYPES_BY_PRODUCTMOMENTS =
  "datahub/G_GET_PRODUCTVARIABLETYPES_BY_PRODUCTMOMENTS";

//APISENSORS

//Mutations
export const M_SET_SENSORS = "apiSensors/M_SET_SENSORS";
export const M_SET_BUOYS = "apiSensors/M_SET_BUOYS";
export const M_SET_STATIONS = "apiSensors/M_SET_STATIONS";
export const M_SET_BUOYTYPES = "apiSensors/M_SET_BUOYTYPES";
export const M_SET_STATIONTYPES = "apiSensors/M_SET_STATIONTYPES";

//Actions
export const A_FETCH_APISENSORS_BUOYS = "apiSensors/A_FETCH_APISENSORS_BUOYS";
export const A_FETCH_APISENSORS_STATIONS =
  "apiSensors/A_FETCH_APISENSORS_STATIONS";
export const A_FETCH_APISENSORS_BUOYTYPES =
  "apiSensors/A_FETCH_APISENSORS_BUOYTYPES";
export const A_FETCH_APISENSORS_STATIONTYPES =
  "apiSensors/A_FETCH_APISENSORS_STATIONTYPES";
export const A_SET_SENSORS = "apiSensors/A_SET_SENSORS";
export const A_SET_CLEAR_SENSORS = "apiSensors/A_SET_CLEAR_SENSORS";

//Getters
export const G_GET_SENSORS_SORTED = "apiSensors/G_GET_SENSORS_SORTED";
export const G_GET_SENSORS_SELECTED = "apiSensors/G_GET_SENSORS_SELECTED";
export const G_GET_SENSORS_COUNT = "apiSensors/G_GET_SENSORS_COUNT";
export const G_GET_BUOYTYPE_NAME = "apiSensors/G_GET_BUOYTYPE_NAME";

//COORDS

//Mutations
export const M_SET_SEARCH_COORDS = "coords/M_SET_SEARCH_COORDS";
export const M_SET_GO_TO_COORDS = "coords/M_SET_GO_TO_COORDS";

//Actions
export const A_SET_SEARCH_COORDS = "coords/A_SET_SEARCH_COORDS";
export const A_SET_GO_TO_COORDS = "coords/A_SET_GO_TO_COORDS";

//FILTERS

//Mutations
export const M_SET_FILTER_MOMENT = "filters/M_SET_FILTER_MOMENT";
export const M_SET_FILTER_VARIABLE = "filters/M_SET_FILTER_VARIABLE";
export const M_SET_FILTER_BUOYSTYPE = "filters/M_SET_FILTER_BUOYSTYPE";
export const M_SET_FILTER_STATIONSTYPE = "filters/M_SET_FILTER_STATIONSTYPE";
export const M_SET_FILTER_ONLY_SPATIAL_FILTERS =
  "filters/M_SET_FILTER_ONLY_SPATIAL_FILTERS";
export const M_SET_FILTER_ALLOW_PRODUCTS =
  "filters/M_SET_FILTER_ALLOW_PRODUCTS";
export const M_SET_FILTER_ALLOW_BUOYS = "filters/M_SET_FILTER_ALLOW_BUOYS";
export const M_SET_FILTER_ALLOW_STATIONS =
  "filters/M_SET_FILTER_ALLOW_STATIONS";

//Actions
export const A_SET_FILTER_MOMENT = "filters/A_SET_FILTER_MOMENT";
export const A_SET_FILTER_VARIABLE = "filters/A_SET_FILTER_VARIABLE";
export const A_SET_FILTER_BUOYSTYPE = "filters/A_SET_FILTER_BUOYSTYPE";
export const A_SET_FILTER_STATIONSTYPE = "filters/A_SET_FILTER_STATIONSTYPE";
export const A_SET_FILTER_ONLY_SPATIAL_FILTERS =
  "filters/A_SET_FILTER_ONLY_SPATIAL_FILTERS";
export const A_SET_FILTER_ALLOW_PRODUCTS = "filters/_SET_FILTER_ALLOW_PRODUCTS";
export const A_SET_FILTER_ALLOW_BUOYS = "filters/A_SET_FILTER_ALLOW_BUOYS";
export const A_SET_FILTER_ALLOW_STATIONS =
  "filters/A_SET_FILTER_ALLOW_STATIONS";

//GETTERS
export const G_GET_FILTERS_READY = "filters/G_GET_FILTERS_READY";
