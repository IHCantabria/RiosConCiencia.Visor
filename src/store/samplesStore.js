import { defineStore } from "pinia";
import {
  getYears,
  getCampaignFilter,
  getDownloadFilters,
} from "@/utils/data-operator-utils.js";
import { CAMPAIGN_VALUES } from "@/config/data-config";
import { ADMIN_ROLE_ID } from "@/config/role-config.js";
import { shallowRef } from "vue";

export const useSamplesStore = defineStore("samplesStore", {
  state: () => {
    return {
      samples: null,
      samplesPicts: null,
      yearOptions: [],
      yearFilter: {},
      campaignOptions: [],
      campaignFilter: {},
      downloadOptions: [],
      downloadFilters: [],
      user: null,
      riverSectionHistoricData: {
        active: false,
        riverSectionId: null,
      },
      riversGeoJson: shallowRef(null),
    };
  },
  getters: {
    getSamplesFiltered() {
      return this.samples
        ? this.samples.filter(
            (sample) =>
              sample.year == this.yearFilter.name &&
              sample.idCampaign == this.campaignFilter.id,
          )
        : [];
    },
    getSamplesFilteredDownloadIds() {
      if (!this.samples) {
        return [];
      } else if (
        !this.downloadFilters.length ||
        (this.downloadFilters.length == 1 && this.downloadFilters[0].id == 0)
      ) {
        return this.samples.map((sample) => sample.id);
      } else {
        return this.samples
          .filter((sample) => this.getFiltersIds.includes(sample.idFilter))
          .map((sample) => sample.id);
      }
    },
    getSamplesPictsFilteredDownloadIds() {
      if (!this.samplesPicts) {
        return [];
      } else if (
        !this.downloadFilters.length ||
        (this.downloadFilters.length == 1 && this.downloadFilters[0].id == 0)
      ) {
        return this.samplesPicts.map((sample) => sample.id);
      } else {
        return this.samplesPicts
          .filter((sample) => this.getFiltersIds.includes(sample.idFilter))
          .map((sample) => sample.id);
      }
    },
    getFiltersIds() {
      return this.downloadFilters.length
        ? this.downloadFilters.map((filter) => filter.id)
        : [];
    },
    getIsFilterAll() {
      return this.downloadFilters[0].id == 0;
    },
    getUserToken() {
      return this.user ? this.user.token : "";
    },
    getUserLogged() {
      return this.user;
    },
    getAdminLogged() {
      return this.user && this.user.roleId == ADMIN_ROLE_ID;
    },
    getSamplesLoaded() {
      return this.samples;
    },
  },
  actions: {
    setSamples(operation) {
      if (!operation) return;
      this.samples = operation;
      const years = getYears(operation);
      const filterCampaign = getCampaignFilter(
        operation,
        years[0],
        CAMPAIGN_VALUES,
      );
      this.yearOptions = years;
      this.yearFilter = years[0];
      this.campaignOptions = CAMPAIGN_VALUES;
      this.campaignFilter = filterCampaign;
      const downloadOptions = getDownloadFilters(years, CAMPAIGN_VALUES);
      this.downloadOptions = downloadOptions;
      this.downloadFilters = [downloadOptions[0]];
    },
    setPictsSamples(operation) {
      this.samplesPicts = operation;
    },
    setRiversGeoJson(operation) {
      this.riversGeoJson = operation;
    },
    setUser(operation) {
      this.user = operation;
    },
    setSampleCampaignFilter(operation) {
      this.campaignFilter = operation;
    },
    setSampleYearFilter(operation) {
      this.yearFilter = operation;
    },
    setSampleDownloadFilters(operation) {
      this.downloadFilters = operation;
    },
    async fetchSamplesAdmin(operation) {
      this.setSamplesAdmin(operation);
    },
    async fetchSamples(operation) {
      this.setSamples(operation);
    },
    async fetchPictsSamples(operation) {
      this.setPictsSamples(operation);
    },
    async fetchUser(operation) {
      this.setUser(operation);
    },
    async setSampleCampaignFilterAction(operation) {
      this.setSampleCampaignFilter(operation);
    },
    async setSampleYearFilterAction(operation) {
      this.setSampleYearFilter(operation);
    },
    async setSampleDownloadFiltersAction(operation) {
      this.setSampleDownloadFilters(operation);
    },
    setRiverSectionHistoricData(idRiverSection) {
      if (!idRiverSection) {
        this.riverSectionHistoricData.active = false;
        this.riverSectionHistoricData.riverSectionId = null;
        return;
      }
      this.riverSectionHistoricData.active = true;
      this.riverSectionHistoricData.riverSectionId = idRiverSection;
    },
  },
});
