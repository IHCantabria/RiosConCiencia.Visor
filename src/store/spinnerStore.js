import { defineStore } from "pinia";
import { spinnerConfig, spinnerEnums } from "@/config/spinner-config";

export const useSpinnerStore = defineStore("spinnerStore", {
  state: () => {
    return {
      mlActive: false,
      calls: 0,
      spinnerConfig: spinnerConfig,
      spinnerEnums: spinnerEnums,
    };
  },
  getters: {
    mlTasksConfig() {
      return this.spinnerConfig?.tasksConfig || null;
    },
  },
  actions: {
    show() {
      this.calls++;
      this.mlActive = true;
    },
    hide() {
      this.calls--;
      if (this.calls == 0) {
        this.mlActive = false;
      }
      if (this.calls == 0) this._clearTasks();
    },
    configureModeSpinner() {
      this.spinnerConfig.mode = this.spinnerEnums.ModesSpinner.progress;
    },
    addTask(name, text) {
      this.spinnerConfig.tasksConfig.tasks.push({
        id: this.spinnerConfig.tasksConfig.tasks.length + 1,
        name,
        text,
        status: this.spinnerEnums.statusTask.pending,
      });
    },
    updateTask(name, status = null, text = null) {
      const indexTask = this.spinnerConfig.tasksConfig.tasks.findIndex(
        (task) => task.name === name,
      );
      if (status) {
        this.spinnerConfig.tasksConfig.tasks[indexTask].status = status;
      }
      if (text) {
        this.spinnerConfig.tasksConfig.tasks[indexTask].text = text;
      }
    },
    _clearTasks() {
      this.spinnerConfig.tasksConfig.tasks = [];
    },
  },
});
