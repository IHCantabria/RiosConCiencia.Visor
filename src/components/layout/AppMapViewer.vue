<template>
  <div class="app-map-viewer">
    <l-map
      :c-center="mapConfigComputed.centerConfig"
      :c-map-default="mapConfigComputed.mapDefaultConfig"
      :c-base-layers="mapConfigComputed.baseLayersConfig"
      :c-switch-layer-base-maps="mapConfigComputed.switchBaseMapsConfig"
      :c-overlay-layers="mapConfigComputed.overlayLayersConfig"
      :c-switch-layer-overlay-maps="mapConfigComputed.switchOverlayMapsConfig"
      :c-markers="markers"
      @toggle-panel="togglePanel"
    ></l-map>
  </div>
</template>
<script>
import LMap from "@/components/ui/LMap.vue";
import { mapGetters } from "vuex";
import * as types from "@/store/types";
import { LEAFLET_CONFIG } from "@/config/leaflet-config.js";
export default {
  components: {
    "l-map": LMap
  },
  data() {
    return {
      leafletConfig: LEAFLET_CONFIG
    };
  },
  computed: {
    ...mapGetters({
      samples: [types.G_GET_SAMPLES_FILTERED]
    }),
    markers() {
      return [...this.samples];
    },
    mapConfigComputed() {
      return {
        centerConfig: this.leafletConfig.centerDefault,
        baseLayersConfig: this.leafletConfig.baseLayersConfig,
        mapDefaultConfig: this.leafletConfig.mapDefaultConfig,
        switchBaseMapsConfig: this.leafletConfig.switchBaseMapsConfig,
        overlayLayersConfig: this.leafletConfig.overlayLayersConfig,
        switchOverlayMapsConfig: this.leafletConfig.switchOverlayMapsConfig
      };
    }
  },
  methods: {
    togglePanel() {
      this.$root.$emit("toggleFilterPanel");
    }
  }
};
</script>
