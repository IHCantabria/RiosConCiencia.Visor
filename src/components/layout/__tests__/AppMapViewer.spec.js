import { shallowMount, createLocalVue } from "@vue/test-utils";
import AppMapViewer from "../AppMapViewer.vue";
import merge from "lodash.merge";
import Vuex from "vuex";
import * as types from "@/store/types";

const localVue = createLocalVue();
localVue.use(Vuex);

function createWrapper(overrides) {
  const defaultMountingOptions = {
    localVue
  };
  return shallowMount(AppMapViewer, merge(defaultMountingOptions, overrides));
}

describe("appMapViewer.vue", () => {
  let store;
  beforeEach(() => {
    store = new Vuex.Store({
      modules: {
        coords: {
          state: {
            goToCoords: null
          },
          actions: {
            [types.A_SET_SEARCH_COORDS]: jest.fn()
          }
        }
      }
    });
  });
  //UNIT TESTS
  it("computed markers update when sensors and products change", () => {
    expect.assertions(1);
    store.hotUpdate({
      getters: {
        [types.G_GET_SENSORS_SELECTED]: () => [{}, {}],
        [types.G_GET_PRODUCTS_SELECTED]: () => [{}, {}]
      }
    });
    const wrapper = createWrapper({
      store
    });
    expect(wrapper.vm.markers).toStrictEqual([{}, {}, {}, {}]);
  });
  it("setSearchCoords is called when searchCoords is called", () => {
    expect.assertions(1);
    const setSearchCoords = jest.fn();
    store.hotUpdate({
      getters: {
        [types.G_GET_SENSORS_SELECTED]: () => [{}, {}],
        [types.G_GET_PRODUCTS_SELECTED]: () => [{}, {}]
      },
      actions: {
        [types.A_SET_SEARCH_COORDS]: setSearchCoords
      }
    });
    const wrapper = createWrapper({
      store
    });
    wrapper.vm.searchCoords("t");
    expect(setSearchCoords).toHaveBeenCalled();
  });
  it("searchExtends is called when setSearchCoords is called", () => {
    expect.assertions(1);
    const setSearchCoords = jest.fn();
    store.hotUpdate({
      getters: {
        [types.G_GET_SENSORS_SELECTED]: () => [{}, {}],
        [types.G_GET_PRODUCTS_SELECTED]: () => [{}, {}]
      },
      actions: {
        [types.A_SET_SEARCH_COORDS]: setSearchCoords
      }
    });
    const wrapper = createWrapper({
      store
    });
    wrapper.vm.searchExtends("t");
    expect(setSearchCoords).toHaveBeenCalled();
  });
  it("setSearchCoords is called when searchGlobal is called", () => {
    expect.assertions(1);
    const setSearchCoords = jest.fn();
    store.hotUpdate({
      getters: {
        [types.G_GET_SENSORS_SELECTED]: () => [{}, {}],
        [types.G_GET_PRODUCTS_SELECTED]: () => [{}, {}]
      },
      actions: {
        [types.A_SET_SEARCH_COORDS]: setSearchCoords
      }
    });
    const wrapper = createWrapper({
      store
    });
    wrapper.vm.searchGlobal("t");
    expect(setSearchCoords).toHaveBeenCalled();
  });
  //SNAPSHOT TESTS
  it("renders correctly", () => {
    expect.assertions(1);
    store.hotUpdate({
      getters: {
        [types.G_GET_SENSORS_SELECTED]: () => [{}, {}],
        [types.G_GET_PRODUCTS_SELECTED]: () => [{}, {}]
      }
    });
    const wrapper = createWrapper({ store });
    expect(wrapper.element).toMatchSnapshot();
  });
});
