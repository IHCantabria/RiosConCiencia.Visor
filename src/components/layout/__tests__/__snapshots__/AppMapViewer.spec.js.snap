// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`appMapViewer.vue renders correctly 1`] = `
<div
  class="app-map-viewer"
>
  <l-map-stub
    cbaselayers="[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]"
    ccenter="[object Object]"
    cdrawoptions="[object Object]"
    cmapdefault="[object Object]"
    cmarkers="[object Object],[object Object],[object Object],[object Object]"
    cswitchlayerbasemaps="[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]"
  />
</div>
`;
